### PR TITLE
#28 Add reusing of cursor, fix potential memory leak with MongoConnection

### PR DESCRIPTION
**Pull request by bdros-g9**

### CODE REVIEW

**Review by BartekGravity — CHANGES_REQUESTED** [review]
Small nitpick about complexity :) Otherwise LGTM

File: src/main/java/com/gravity9/mongocse/DefaultMongoClientProvider.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+	private MongoClient client;
+
+    DefaultMongoClientProvider(String mongoConnectionUri) {
+        this.mongoConnectionUri = mongoConnectionUri;
+    }
+
+	@Override
+    public MongoClient getClient() {
+		if (client == null) {
```
Comment (piotrlg9) [inline src/main/java/com/gravity9/mongocse/DefaultMongoClientProvider.java:25]: this part is not thread-safe, maybe LazyHolder?
* https://stackoverflow.com/a/8297830

File: src/main/java/com/gravity9/mongocse/MongoCseManager.java [hunk 2]
```diff
@@ -39,20 +44,21 @@ private Map<Integer, MongoChangeStreamWorker> createWorkers() {
             throw new IllegalArgumentException("Cannot initialize with less than 1 partition!");
         }
 
-		var collectionName = clusterConfig.getCollection();
-		Map<Integer, MongoChangeStreamWorker> workersByPartitionMap = new HashMap<>();
+        var collectionName = clusterConfig.getCollection();
+        Map<Integer, MongoChangeStreamWorker> workersByPartitionMap = new HashMap<>();
```
Comment (piotrlg9) [inline src/main/java/com/gravity9/mongocse/MongoCseManager.java:48]: initial capacity for map can be provided here

File: src/main/java/com/gravity9/mongocse/MongoChangeStreamWorker.java [hunk 3]
```diff
@@ -115,73 +103,74 @@ public void run() {
...
+                } while (isReadingFromChangeStream);
             } catch (Exception ex) {
                 log.error("Exception while processing change", ex);
-            } finally {
-                if (!isReadingFromChangeStream) {
-                    shutdownLatch.countDown();
-                    LoggingUtil.removeLoggingContext();
-                }
             }
         } while (isReadingFromChangeStream);
```
Comment (piotrlg9) [inline src/main/java/com/gravity9/mongocse/MongoChangeStreamWorker.java:151]: isReadingFromChangeStream is not set to `false` anywhere(?)

File: src/main/java/com/gravity9/mongocse/MongoChangeStreamWorker.java [hunk 4]
```diff
@@ -115,73 +103,74 @@ public void run() {
...
-                        case DELETE:
-                            log.info("DELETE, document id: {}", changedDocumentId);
-                            break;
-                        default:
-                            if (canLogSensitiveData()) {
-                                log.debug("{} document: {}", document.getOperationType().name(), toJson(document.getFullDocument()));
-                            } else {
-                                log.info("{} document id: {}", document.getOperationType().name(), changedDocumentId);
-                            }
+                do {
```
[the remark this answers is not in the document]
Comment (bdros-g9) [reply src/main/java/com/gravity9/mongocse/MongoChangeStreamWorker.java:102]: I made small code refactor

File: src/main/java/com/gravity9/mongocse/DefaultMongoClientProvider.java (hunk 1 above)
Comment (bdros-g9) [reply src/main/java/com/gravity9/mongocse/DefaultMongoClientProvider.java:25]: I preferred synchronization on provider instance, to ensure both cration and destruction is synchronized as well.

File: src/main/java/com/gravity9/mongocse/DefaultMongoClientProvider.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+			.applyConnectionString(connectionString)
+			.codecRegistry(codecRegistry)
+			.build();
+
+		return MongoClients.create(clientSettings);
+	}
+
+	@Override
+	public void close() {
+		if (client != null) {
```
[the remark this answers is not in the document]
Comment (bdros-g9) [reply src/main/java/com/gravity9/mongocse/DefaultMongoClientProvider.java:51]: Like in previous comment

**Review by BartekGravity — APPROVED** [review]
LGTM

**Pull request merged**